### ISSUE TITLE
hub api queries should follow redirects

### ISSUE THREAD

**Issue by mlewand** [github/hub#2449]
**The problem I'm trying to solve:**

It seems that the `hub api` (using v3 REST API) command does not follow redirects.

I used in the past following command `hub api repos/orgName/repoName/hooks --method "POST" --input /query-fields.json` in the past which I recall was working fine.

But recently when I used it I started to get `{"message":"Moved Permanently","url":"https://api.github.com/repositories/123456789/hooks","documentation_url":"https://developer.github.com/v3/#http-redirects"}` responses in the CLI and the API call does not succeed.

When testing this URL in the browser, the browser is redirected correctly so I assume that all the headers are fine.

I've seen some redirect action published in recent 2.14 release, maybe its related? Or maybe simply GH changed something in the API?

Fetching repo id is problematic if you just want to do one or two quick REST api queries.

**How I imagine hub could expose this functionality:**

Basically I'd expect that after getting answer like that it uses `response.url` to request the query once again there.

**Version I'm using:**

```
git version 2.25.0.windows.1
hub version 2.14.0
```

**Comment by mlewand** [comment]
Interestingly, I tried this API call on other repositories and it works...

Edit: I thought it was a random thing, but the reason for this to happen was that this particular repository was renamed in the past 😂 and the command contained the former name.

So the issue priority is lower I would think, but that was tricky to figure out.

**Comment by mislav** [comment]
Hi thank you for writing in! You are right that this change was intentional in 2.14, as per release notes.

> I used in the past following command `hub api repos/orgName/repoName/hooks --method "POST" --input /query-fields.json` in the past which I recall was working fine.

Prior to this release, it might have _not_ been working fine, but there was no error thrown so you might not have noticed.

GitHub will redirect renamed repositories to the new name using HTTP 301 redirects. However, when Go's `net/http` library follows these redirects for POST requests (or any non-GET request), it will convert them to a GET. Therefore, your request would be converted to a `GET repos/orgName/repoName/hooks`, which would technically _succeed_ since that endpoint also exists, but the original POST body you've intended was dropped. This could happen without you noticing.

There is no easy way to teach Go to keep the method as POST when following 301/302, so instead I've opted to issue an error instead: https://github.com/github/hub/commit/ff6b131d93b3bf8c3dd8a1b60fc3536ef87f6e1f. The server should ideally respond with 308/307 to non-GET requests if it wanted clients to follow them without changing the HTTP method. If GitHub API doesn't do this, it might be a bug and I encourage you to report it to our Support!

In the meantime, the way you work around it is to use the latest `orgName/repoName` after the repository was renamed.

**Comment by mlewand** [comment]
Thanks for clarifying this 👍
